### PR TITLE
[IMP] web: can expand form view dialog to full form view

### PR DESCRIPTION
**Pull request by SimonGenin**
This task allows to expand a form view that is inside a dialog to a
full form view. It is possible when it is using a FormViewDialog, or
when using a x2ManyX2ManyFieldDialog. For the later, there is a
limitation: you cannot expand a newly created x2many record if it has
just been created and the root record has not been saved yet.
task id 4080265

### CODE REVIEW

File: addons/web/static/src/core/dialog/dialog.scss [hunk 1]
```diff
@@ -98,3 +98,10 @@
         }
     }
 }
+
+.btn-expand {
```
Comment (aab-odoo) [inline addons/web/static/src/core/dialog/dialog.scss:102]: on top
```scss
    .modal-header .o_expand_button {
        ...
    }
```

File: addons/web/static/src/core/dialog/dialog.xml [hunk 2]
```diff
@@ -37,9 +38,12 @@
         <t t-if="fullscreen">
             <button class="btn oi oi-arrow-left" data-bs-dismiss="modal" aria-label="Close" t-on-click="dismiss" />
         </t>
-        <h4 class="modal-title text-break" t-att-class="{ 'me-auto': fullscreen }">
+        <h4 class="modal-title text-break flex-grow-1" t-att-class="{ 'me-auto': fullscreen }">
             <t t-esc="props.title"/>
         </h4>
+        <t t-if="!fullscreen and onExpand">
+            <button type="button" class="fa fa-expand btn btn-expand" aria-label="Expand" tabindex="-1" t-on-click="onExpand"></button>
```
Comment (aab-odoo) [inline addons/web/static/src/core/dialog/dialog.xml:45]:
```suggestion
            <button type="button" class="fa fa-expand btn btn-expand" aria-label="Expand" tabindex="-1" t-on-click="onExpand"/>
```

File: addons/web/static/src/views/fields/relational_utils.js [hunk 3]
```diff
@@ -595,6 +605,19 @@ export class X2ManyFieldDialog extends Component {
...
 
+    get dialogProps() {
+        const props = {
+            title: this.title,
+            withBodyPadding: false,
+            modalRef: this.modalRef,
+            contentClass: this.contentClass,
+        };
+        if (!this.record.isNew) {
+            props.onExpand = this.onDialogExpand;
```
Comment (aab-odoo) [inline addons/web/static/src/views/fields/relational_utils.js:616]: inline

**Review by aab-odoo — COMMENTED** [review]
add tests:
 - x2many (existing record)
 - x2many (new record)
 - formview dialog (new)
 - formview dialog (existing)
 - formview dialog  => save and new
 - mobile (no button)

File: addons/web/static/tests/views/fields/one2many_field.test.js [hunk 4]
```diff
@@ -12868,3 +12868,57 @@ test("onchange on x2many add and delete x2m record, returning to initial state",
...
+    expect(".o_field_widget[name=turtles] .o_data_row").toHaveCount(1);
+    await contains(queryFirst(".o_field_widget[name=turtles] .o_data_cell")).click();
+    expect(".o_dialog .o_form_view").toHaveCount(1);
+    expect(".o_dialog .modal-header .o_expand_button").toHaveCount(0);
+    await contains(".o_dialog .modal-header .btn-close").click();
+    await clickSave();
+    await contains(queryLast(".o_field_widget[name=turtles] .o_data_cell")).click();
+    expect(".o_dialog .o_form_view").toHaveCount(1);
+    expect(".o_dialog .modal-header .o_expand_button").toHaveCount(1);
+});
```
Comment (SimonGenin) [inline addons/web/static/tests/views/fields/one2many_field.test.js:12924]: not sure this one really brings anything new. As the other test requires the existing model to be saved for the newly created one2many to be expandable

Comment (aab-odoo) [reply addons/web/static/tests/views/fields/one2many_field.test.js:12924]: Indeed, I would keep only the first one. However, you only assert the presence of the button, but you don't click on it (to assert that the expected action is executed, with a mockService("action"))

File: addons/web/static/tests/views/view_dialogs/form_view_dialog.test.js [hunk 5]
```diff
@@ -382,3 +384,114 @@ test("Save a FormViewDialog when a required field is empty don't close the dialo
     await animationFrame();
     expect(".modal").toHaveCount(0, { message: "modal should be closed" });
 });
+
+test.tags("desktop")("new record has an expand button", async () => {
+    Partner._views["form,false"] = /* xml */ `<form><field name="foo"/></form>`;
+    const numberOfRecordsBeforeSave = Partner._records.length;
```
Comment (aab-odoo) [inline addons/web/static/tests/views/view_dialogs/form_view_dialog.test.js:390]: I would erase Partner._records here (like you do in the last test, but with empty list here), s.t. you have full control, and you can assert that the id is 1.

**Review by aab-odoo — APPROVED** [review]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo r+

**Pull request closed without merging**